### ISSUE TITLE
BrowseEverything S3 setup

### ISSUE THREAD

**Issue by crisr15** [scientist-softserv/britishlibrary#173]

**Comment by jeremyf** [comment]
Rory pointed me to https://github.com/research-technologies/browse-everything/tree/sharepoint_provider, a fork that includes the SharePoint work.  We'll need to discuss if this is pushed back to Samvera or if we proceed with this fork.

**Comment by jeremyf** [comment]
On 2022-12-07 I reached out to Jenny [via Slack](https://assaydepot.slack.com/archives/G0313NK2LJ0/p1670426017657089) regarding credentials.  There is ongoing conversations via email with their IT department concerning BrowseEverything and it's implementation.

**Comment by j-basford** [comment]
BL is OK with the Sharepoint access, but are upgrading all Sharepoint estate and are busy with that. No Technology resource til Q2 2023 so suggest we continue with S3 only and come back to SHarepoint at a later date (there may be a separate ticket for Sharepoint BE)

**Comment by j-basford** [comment]
To proceed with this as just with S3

**Comment by ShanaLMoore** [comment]
@j-basford where can we get credentials to your s3?

**Comment by j-basford** [comment]
> 

@cziaarm should be able to provide these.

**Comment by ShanaLMoore** [comment]
cc @jillpe moving this out of the sprint: waiting on client decision.

The client needs to have policy and other prior discussions before implementation is done. We will mark this as a blocker until further notice.

**Comment by j-basford** [comment]
We cannot use a BL S3 - can we use a CoSector one to confirm this works and then we will revisit after development is verified @jillpe @cziaarm

**Comment by cziaarm** [comment]
Have created bucket (and a user with access keys). May need some guidance on what to add to browse_everything.yml

**Comment by ShanaLMoore** [comment]
> Have created bucket (and a user with access keys). May need some guidance on what to add to browse_everything.yml

Hi @cziaarm [This](https://github.com/samvera/browse-everything/blob/main/lib/generators/browse_everything/templates/browse_everything_providers.yml.example) example may help. Hopefully it's a plug and play type of set up.

**Comment by cziaarm** [comment]
Hi @ShanaLMoore 

I have what may a useful set of value in place in the BE yaml. I'm getting an odd error that makes me feel like BE has gone wrong in constructing a url along the way (perhaps because of misconfiguration). The config looks like this:

```
s3:
  bucket: temp-bl-bucket-for-browse-everything.s3.amazonaws.com
  app_key: [MY_APP_KEY]
  app_secret: [MY_APP_SECRET]
  region: eu-west-1   
```

I've left out the `response_type` and `expires_in` options, as there is no mention of those in the docs.

I've checked out these value with a simple get via postman and they are all good. 

In hyku, I'm getting the s3 option and the modal that looks like this:

![Image](https://user-images.githubusercontent.com/1931982/224346357-362f4f29-d2bb-435a-9bed-93d7db0caa56.png)

but when I click on "connect" I end up with an error... I don't think the error is relevant directly as I think it is the URL that is at fault:

`http://bl.bl.test/concern/articles/&state=s3` 

Is the url that I'm ending up at and unsurprisingly causing an error

On closer inspection I can see the link for the "connect" button is:

`<a class="btn btn-primary ev-auth" target="blank" id="provider_auth" href="&state=s3">Connect to S3</a>`

So it feels like BE is missing something important here?

**Comment by cziaarm** [comment]
So link is generated by `auth_link` which is not overridden fo the s3 provider, hence the lack of complete url... but I suspect that would be a link to authenticate a user for the provider. In this case my bucket has key access and that key/secret is in the config, so no need for an auth step? I'd love for someone to show me round a working s3 example, as they must exist, and I'm obviously doing something wrong

**Comment by jillpe** [comment]
Hi @cziaarm! Shana is out today, but our team has some documentation they wrote that might be helpful:

[Adding Browse Everything to a Hyrax Application
](https://docs.google.com/document/d/1K8mG2ONEm_X3PLoMg0L8_ZXkANbx-18MJeZM_B5Vkw0/edit)
[Server Side Storage Support Setup](https://docs.google.com/document/d/1CyJLkt0KDfs6z8lSYvEoyiIPGQNqSIugbbunZPnP_ag/edit#heading=h.vvicl0luiabh)

I'm also trying to find someone on the team who has experience with this and could pair with you

**Comment by cziaarm** [comment]
@NoraRamsey @grahamjevon @j-basford 

The S3 provider has been configured and is now available on the staging repository. You will need to configure a desktop client to be able to put things into the bucket. I have used Winscp (It is a Windows SCP/FTP client that understands S3). If you find your preferred file transfer desktop client that can use the AWSS3 protocol, then I will be able to share the access keys with you and you'll be able to use the "Add Cloud Files" feature both in the normal upload workflow and the Bulkrax import.

I'll be on slack

**Comment by cziaarm** [comment]
Hi Rory, regarding BE, I uploaded a 2.5GB file today to S3. The work has appeared in the repo, but the file has yet to load in the repo. Is it possible to see if that is still loading behind the scenes or does this indicate an issue? I ran the upload a few hours ago and got the familiar Chrome error.
Incidentally, the work was duplicated (so two copies of the work have appeared and two copies of the importer are showing). I'm not sure if this was a human error (perhaps I double clicked import) or if the this was technical error. I thought I'd wait until we knew if the import was still running behind the scenes before testing this again.

**Comment by grahamjevon** [comment]
[Importer ](https://bl.bl-staging.notch8.cloud/importers/121?locale=en)successfully imported [work with 594MB file](https://bl.bl-staging.notch8.cloud/concern/articles/cf78b1fd-b829-4b0f-9eea-4533b023cc97?locale=en) using BE. Everything happened as expected.

Importer with 2GB file resulted in a "504 Gateway Time-Out nginx". This message appeared about 1-2 minute safter starting the importer. When I went to the Importer history, there were two duplicate importers for this, which both said complete:

https://bl.bl-staging.notch8.cloud/importers/122
https://bl.bl-staging.notch8.cloud/importers/123

This resulted in two works being created:

https://bl.bl-staging.notch8.cloud/concern/articles/3f2955a1-3c36-4784-a913-dc7e2790142c
https://bl.bl-staging.notch8.cloud/concern/articles/f30f3791-970a-4ddb-85ac-8931b7264159

While the filename appears in the items list, the item has no file size and it cannot be downloaded. This suggests that the upload of the file failed. This seems to replicate my experience when testing BE back in July.

**Comment by cziaarm** [comment]
Key here I think is that when we use BE via Bulkrax the _web_ does the download. This is different to when we use BE in the upload context. In that case the _worker_ asynchronously imports the S3 URL and then it is attached to the file_set.
